[GPU] DeformableConvolution-8

ytorzuk-altran opened the pull request:
### Details:
 - *DeformableConvolution-8 reference kernel*
 - *DeformableConvolution-8 optimized kernel*
 - *Single layer tests which cover DeformableConvolution-8 functionality*

lznamens commented on inference-engine/tests/functional/plugin/gpu/shared_tests_instances/single_layer_tests/deformable_convolution.cpp:
```diff
@@ -0,0 +1,157 @@
+    ::testing::ValuesIn(with_bilinear_interpolation_pad),
+    ::testing::ValuesIn(with_modulated_scalar)
+);
+//
```
Please add comment or remove this unnecessary code.

ytorzuk-altran replied:
done

lznamens commented on inference-engine/thirdparty/clDNN/api/cldnn/primitives/convolution.hpp:
```diff
@@ -830,6 +894,30 @@ struct deformable_interp : public primitive_base<deformable_interp> {
               padding_below(tensor(0)) {}
 
+    deformable_interp(const primitive_id& id,
+                          const std::vector<primitive_id>& inputs,
```
nit: alignment

ytorzuk-altran replied:
done

lznamens commented on inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/deformable_convolution_gpu_bfyx_ref.cl:
```diff
@@ -48,26 +51,33 @@ KERNEL(deformable_convolution_gpu_bfyx_ref)(
+                const bool y_is_out_of_boundaries = transformed_y >= INPUT0_SIZE_Y || transformed_y < 0;
+#endif
+                if(!y_is_out_of_boundaries) {
+                    if(!x_is_out_of_boundaries) {
```
Maybe it makes sence to combine both conditions into one with "&"?

ytorzuk-altran replied:
done

lznamens reviewed (commented):
In general this PR looks good to me, please resolve minor comments.

sshlyapn commented on inference-engine/src/cldnn_engine/ops/convolution.cpp:
```diff
@@ -303,6 +303,78 @@ static void CreateDeformableConvolutionOp(Program& p, const std::shared_ptr<ngra
+    std::vector<cldnn::primitive_id> weights = {inputs[2]};
+    // Remove weights from inputs
+    inputs.erase(inputs.begin() + 2);
+    if (params.groups == 1 && op->get_input_element_type(0).bitwidth() == 32) {
```
Why do we have such limitations for `bitwidth`? Does it really work let's say for `int32`? Otherwise it should be comparation with something like `ngraph::element::Type_t::f32`

ytorzuk-altran replied:
The original implementation failed on some of SLTs I've added in fp16 precision. I drilled down into the problem and figured out that it's related to fp16 precision math in the kernel, which was not enough for deformable offsets calculation, so that neighbor feature value was calculated instead the correct one. I fixed the problem adding casting of deformable offset value to fp32 in the new patch set, therefore the limitation for optimized deformable convolution implementation was eliminated.

sshlyapn commented on inference-engine/src/cldnn_engine/ops/convolution.cpp:
```diff
@@ -303,6 +303,78 @@ static void CreateDeformableConvolutionOp(Program& p, const std::shared_ptr<ngra
     }
 }
 
+static void CreateDeformableConvolutionOp(Program& p, const std::shared_ptr<ngraph::op::v8::DeformableConvolution>& op) {
```
Can we extract generic code from old `CreateDeformableConvolutionOp(...)` and new functions into some common function `CreateDeformableConvolutionCommon`?

ytorzuk-altran replied:
done

ytorzuk-altran commented on inference-engine/src/cldnn_engine/ops/convolution.cpp:
```diff
@@ -303,6 +303,78 @@ static void CreateDeformableConvolutionOp(Program& p, const std::shared_ptr<ngra
+                CldnnTensorFromIEDims(outDims),
+                kernel,
+                op->get_bilinear_interpolation_pad(),
+                op->get_friendly_name());
```
done

ytorzuk-altran commented on inference-engine/src/cldnn_engine/ops/convolution.cpp:
```diff
@@ -303,6 +303,78 @@ static void CreateDeformableConvolutionOp(Program& p, const std::shared_ptr<ngra
+        p.AddPrimitive(defConvPrimInterp);
+        p.AddInnerPrimitiveToProfiler(defConvLayerNameInterp, defConvLayerNameConv, op);
+        auto defConvPrim = cldnn::deformable_conv(defConvLayerNameConv,
+                defConvLayerNameInterp,
```
done

ytorzuk-altran commented on inference-engine/src/cldnn_engine/ops/convolution.cpp:
```diff
@@ -303,6 +303,78 @@ static void CreateDeformableConvolutionOp(Program& p, const std::shared_ptr<ngra
+                    cldnn::feature(1),
+                    cldnn::spatial(weights_shape[nonSpatialDimsNum + 2],
+                            weights_shape[nonSpatialDimsNum + 1],
+                            weights_shape[nonSpatialDimsNum + 0]));
```
done

ytorzuk-altran commented on inference-engine/src/cldnn_engine/ops/convolution.cpp:
```diff
@@ -303,6 +303,78 @@ static void CreateDeformableConvolutionOp(Program& p, const std::shared_ptr<ngra
+                    cldnn::feature(1),
+                    cldnn::spatial(weights_shape[nonSpatialDimsNum + 1],
+                            weights_shape[nonSpatialDimsNum + 0],
+                            1));
```
done

ytorzuk-altran commented on inference-engine/src/cldnn_engine/ops/convolution.cpp:
```diff
@@ -303,6 +303,78 @@ static void CreateDeformableConvolutionOp(Program& p, const std::shared_ptr<ngra
+        p.AddPrimitiveToProfiler(defConvLayerNameConv, op);
+    } else {
+        auto convPrim = cldnn::convolution(layerName,
+                inputs,
```
done

ytorzuk-altran commented on inference-engine/thirdparty/clDNN/api/cldnn/primitives/convolution.hpp:
```diff
@@ -830,6 +894,30 @@ struct deformable_interp : public primitive_base<deformable_interp> {
               padding_above(tensor(0)),
               padding_below(tensor(0)) {}
 
+    deformable_interp(const primitive_id& id,
```
done

ytorzuk-altran commented:
@lznamens, @sshlyapn I addressed all your comments and rebased onto the latest master. @lznamens Did you use a tool for proper formatting? If so, could you please share a config for this tool?

lznamens commented:
> @lznamens, @sshlyapn I addressed all your comments and rebased onto the latest master. @lznamens Did you use a tool for proper formatting? If so, could you please share a config for this tool?

I use VS-code with Tab size = 4 setting + git diff for common checks (e.g. whitespaces). I don't use special tools for text formatting.

lznamens commented on src/plugins/intel_gpu/src/plugin/ops/convolution.cpp:
```diff
@@ -233,61 +233,48 @@ static void CreateGroupConvolutionBackpropDataOp(Program& p, const std::shared_p
+                    cldnn::feature(1),
+                    cldnn::spatial(weights_shape[nonSpatialDimsNum + 1],
+                                   weights_shape[nonSpatialDimsNum + 0],
+                                   1));
```
I suppose that code alignment in these 4 lines was correct before your changes related to variable renaming.

ytorzuk-altran replied:
done

lznamens commented on inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/deformable_convolution_gpu_bfyx_ref.cl:
```diff
@@ -48,53 +51,74 @@ KERNEL(deformable_convolution_gpu_bfyx_ref)(
+                        dotProd += value * weights[filter_idx];
+#endif
+                    }
+                } // !y_is_out_of_boundaries && !x_is_out_of_boundaries
```
I suppose it makes sence to remove empty 123 line.

ytorzuk-altran replied:
done

lznamens commented on inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/deformable_convolution_gpu_bfyx_ref.cl:
```diff
@@ -48,53 +51,74 @@ KERNEL(deformable_convolution_gpu_bfyx_ref)(
+                        INPUT0_TYPE top_left     = top_y_index < 0 || left_x_index < 0 ? 0 : (INPUT0_TYPE)data[GET_DATA_INDEX(INPUT0, b, ifm, top_y_index, left_x_index) + in_split_offset];
+                        INPUT0_TYPE top_right    = top_y_index < 0 || right_x_index >= INPUT0_SIZE_X ? 0 :
+                            (INPUT0_TYPE)data[GET_DATA_INDEX(INPUT0, b, ifm, top_y_index, right_x_index) + in_split_offset];
+                        INPUT0_TYPE bottom_left  = bottom_y_index >= INPUT0_SIZE_Y || left_x_index < 0 ? 0 :
+                            (INPUT0_TYPE)data[GET_DATA_INDEX(INPUT0, b, ifm, bottom_y_index, left_x_index) + in_split_offset];
+                        INPUT0_TYPE bottom_right = bottom_y_index >= INPUT0_SIZE_Y || right_x_index >= INPUT0_SIZE_X ? 0 :
+                                (INPUT0_TYPE)data[GET_DATA_INDEX(INPUT0, b, ifm, bottom_y_index, right_x_index) + in_split_offset];
```
```suggestion
                        INPUT0_TYPE bottom_left  = bottom_y_index >= INPUT0_SIZE_Y || left_x_index < 0 ? 0 :
                            (INPUT0_TYPE)data[GET_DATA_INDEX(INPUT0, b, ifm, bottom_y_index, left_x_index) + in_split_offset];
                        INPUT0_TYPE bottom_right = bottom_y_index >= INPUT0_SIZE_Y || right_x_index >= INPUT0_SIZE_X ? 0 :
                            (INPUT0_TYPE)data[GET_DATA_INDEX(INPUT0, b, ifm, bottom_y_index, right_x_index) + in_split_offset];
```

ytorzuk-altran replied:
done

lznamens commented on inference-engine/thirdparty/clDNN/kernel_selector/core/cl_kernels/deformable_convolution_gpu_bfyx_ref.cl:
```diff
@@ -48,53 +51,74 @@ KERNEL(deformable_convolution_gpu_bfyx_ref)(
+                const bool x_is_out_of_boundaries = transformed_x >= INPUT0_SIZE_X || transformed_x < 0;
+                const bool y_is_out_of_boundaries = transformed_y >= INPUT0_SIZE_Y || transformed_y < 0;
+#endif
+                if(!y_is_out_of_boundaries && !x_is_out_of_boundaries) {
```
```suggestion
                if (!y_is_out_of_boundaries && !x_is_out_of_boundaries) {
```

ytorzuk-altran replied:
done

vladimir-paramuzov commented on inference-engine/thirdparty/clDNN/api/cldnn/primitives/convolution.hpp:
```diff
@@ -774,7 +764,12 @@ struct convolution : public primitive_base<convolution> {
+    /// then bilinear interpolation is performed on the zero padded feature map.
+    /// If bilinear_interpolation_pad is false and the sampling location is within one pixel outside of the feature map boundary,
+    /// then the sampling location shifts to the inner boundary of the feature map.
+    bool bilinear_interpolation_pad;
```
Seems like this field may be uninitialized in some c-tors.
`bool bilinear_interpolation_pad = false`

ytorzuk-altran replied:
Added `bilinear_interpolation_pad` initialization

vladimir-paramuzov commented on inference-engine/thirdparty/clDNN/kernel_selector/core/actual_kernels/convolution/convolution_params.h:
```diff
@@ -4,6 +4,7 @@
 
 #pragma once
 
+#include "cldnn/primitives/convolution.hpp"
```
Is it really needed?

ytorzuk-altran replied:
It's a leftover from the 1st patch set of the PR which is not need anymore. Dropped.

ytorzuk-altran commented:
@lznamens @vladimir-paramuzov All your comments were addressed. Changes are rebased onto the latest master.

lznamens reviewed (approved):
LGTM.
